converter/geant4/parser.py

```diff
@@ -0,0 +1,211 @@
+        })
+
+        ET.SubElement(gdml_root, "define")
+        used_materials: Set[str] = set()
+        self._collect_materials(world, used_materials)
+
+        materials_xml = ET.SubElement(gdml_root, "materials")
+        if "BlackHole" in used_materials:
+            mat = ET.SubElement(materials_xml, "material", {"name": "BlackHole", "state": "solid"})
+            ET.SubElement(mat, "T", {"unit": "K", "value": "293.15"})
+            ET.SubElement(mat, "MEE", {"unit": "eV", "value": "823"})
+            ET.SubElement(mat, "D", {"unit": "g/cm3", "value": "1e+8"})
+            ET.SubElement(mat, "fraction", {"n": "1", "ref": "G4_Pb"})
+
+        solids_xml = ET.SubElement(gdml_root, "solids")
+        structure_xml = ET.SubElement(gdml_root, "structure")
+        name_counters = {"solid": {}, "logic": {}, "phys": {}}
+
+        self._emit_node_postorder(world, solids_xml, structure_xml, name_counters)
+
+        setup_xml = ET.SubElement(gdml_root, "setup", {"name": "Default", "version": "1.0"})
+        world_logic_name = self._choose_logic_name(world)
+        ET.SubElement(setup_xml, "world", {"ref": world_logic_name})
+
+        return self._prettify_xml(gdml_root)
+
+    def _generate_empty_gdml(self) -> str:
+        """Generate empty GDML from gemetry node"""
+        root = ET.Element("gdml", {
+            "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
+            "xsi:noNamespaceSchemaLocation": "http://cern.ch/service-spi/app/releases/GDML/schema/gdml.xsd",
+        })
+        ET.SubElement(root, "define")
+        ET.SubElement(root, "materials")
+        solids = ET.SubElement(root, "solids")
+        ET.SubElement(solids, "box", {
+            "name": "solidWorld", "x": _to_mm_str(100), "y": _to_mm_str(100), "z": _to_mm_str(100), "lunit": "mm"
+        })
+        structure = ET.SubElement(root, "structure")
+        vol = ET.SubElement(structure, "volume", {"name": "logicWorld"})
+        ET.SubElement(vol, "materialref", {"ref": "G4_Galactic"})
+        ET.SubElement(vol, "solidref", {"ref": "solidWorld"})
+        setup = ET.SubElement(root, "setup", {"name": "Default", "version": "1.0"})
+        ET.SubElement(setup, "world", {"ref": "logicWorld"})
+
+        return self._prettify_xml(root)
+
+    def _collect_materials(self, node: dict, acc: Set[str]) -> None:
+        """Collect all material nodes recursively"""
+        mat = node.get("simulationMaterial", {})
+        g4 = mat.get("geant4_name")
+        if g4:
+            acc.add(g4)
+        for ch in node.get("children", []):
+            self._collect_materials(ch, acc)
+
+    @staticmethod
+    def _unique_name(base: str, kind: str, counters: Dict[str, Dict[str, int]]) -> str:
+        """Generate a unique name for a given base string and category (solid, logic, phys)."""
+        used = counters[kind]
+        if base not in used:
+            used[base] = 1
+            return base
+        used[base] += 1
+        return f"{base}{used[base]}"
```

Comment:
That method is not using any of class stuff, maybe move it simply to a module level function, instead of keeping it class static method ?